core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/sendsemantics/partitioners/ReshapePartitioner.scala

```diff
@@ -17,10 +17,14 @@ import scala.collection.mutable.ArrayBuffer
   */
 class ReshapePartitioner(partitioner: Partitioner) extends Partitioner {
 
-  override def getBucketIndex(tuple: ITuple): Int = {
-    val bucketIndex = partitioner.getBucketIndex(tuple)
-    val newDestReceiver = recordSampleAndGetReceiverForReshape(bucketIndex)
-    originalReceiverIndexMapping(newDestReceiver)
+  override def getBucketIndex(tuple: ITuple): Iterator[Int] = {
+    val it = partitioner.getBucketIndex(tuple)
+    val itMapping = Iterator()
+    for (bucketIndex <- it) {
+      val newDestReceiver = recordSampleAndGetReceiverForReshape(bucketIndex)
+      itMapping ++ Iterator(originalReceiverIndexMapping(newDestReceiver))
+    }
+    it
```

Comment:
Simplify this with some functional methods (e.g., toIterator, map, reduce, etc).